File: pharmacy-api/shared/models/database_models.go
package models

import (
	"database/sql"
	"time"
)

// MYSQL or GORM moment idk, on Postgres all alright
type WorkTime struct {
	ID        uint           `gorm:"primaryKey"`
	Date      sql.NullString `gorm:"type:DATE; not null "`
	IsOpened  bool           `gorm:"not null"`
	TimeStart sql.NullString `gorm:"type:TIME"`
	TimeEnd   sql.NullString `gorm:"type:TIME"`
}

type Producer struct {
	ID           uint    `gorm:"primaryKey"`
	ProducerName string  `gorm:"size:40; not null"`
	Goods        []Goods `gorm:"foreignKey:ProducerID"`
}

type Tag struct {
	ID      uint    `gorm:"primaryKey"`
	TagName string  `gorm:"size:20; not null"`
	Goods   []Goods `gorm:"many2many:goods_tags"`
}

type GoodsOrders struct {
	OrderID  uint  `gorm:"primaryKey"`
	GoodsID  uint  `gorm:"primaryKey"`
	Quantity uint  `gorm:"not null"`
	Order    Order `gorm:"foreignKey:OrderID"`
	Goods    Goods `gorm:"foreignKey:GoodsID"`
}

type Order struct {
	ID          uint          `gorm:"primaryKey"`
	ClientFIO   string        `gorm:"size:30; not null"`
	ClientEmail string        `gorm:"size:30"`
	ClientPhone string        `gorm:"size:18; not null"`
	Goods       []GoodsOrders `gorm:"foreignKey:OrderID; constraint:OnDelete:CASCADE"`
}

type Goods struct {
	ID                   uint          `gorm:"primaryKey"`
	Name                 string        `gorm:"size:64; not null"`
	Image                string        `gorm:"size:64; not null"`
	ProducerID           uint          `gorm:"index; not null"`
	Producer             Producer      `gorm:"foreignKey:ProducerID; not null"`
	IsInStock            bool          `gorm:"not null"`
	Tags                 []Tag         `gorm:"many2many:goods_tags; not null"`
	Orders               []GoodsOrders `gorm:"foreignKey:GoodsID"`
	Instruction          string        `gorm:"size:1024"`
	Description          string        `gorm:"size:1024"`
	IsPrescriptionNeeded bool          `gorm:"not null"`
	Price                uint          `gorm:"not null"`
}

type Announcement struct {
	ID       uint      `gorm:"primaryKey"`
	DateTime time.Time `gorm:"not null"`
	From     string    `gorm:"size:64; not null"`
	Announce string    `gorm:"size:2048; not null"`
}

type User struct {
	ID           uint   `gorm:"primaryKey"`
	Login        string `gorm:"size:40; not null; unique; index"`
	UserName     string `gorm:"size:40; not null"`
	RoleID       uint   `gorm:"index; not null"`
	Role         Role   `gorm:"foreignKey:RoleID; not null"`
	PasswordHash []byte `gorm:"size:128; not null"`
}

type Role struct {
	ID          uint         `gorm:"primaryKey"`
	Name        string       `gorm:"size:20; not null"`
	User        []User       `gorm:"foreignKey:RoleID"`
	Permissions []Permission `gorm:"many2many:role_permission; not null; constraint:OnDelete:CASCADE"`
}

type Permission struct {
	ID     uint   `gorm:"primaryKey"`
	Action string `gorm:"size:20; not null"`
	Role   []Role `gorm:"many2many:role_permission"`
}

File: pharmacy-api/src/shared/models/database_models.ts
import { Entity, PrimaryGeneratedColumn, Column, ManyToOne, OneToMany, ManyToMany, JoinTable, JoinColumn, DataSource } from 'typeorm';
import { Pharmacy1776354362570 } from '@pharmacy/src/migration/1776354362570-pharmacy';

@Entity('schedules')
export class Schedule {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: 'date', nullable: true })
  date: Date | null;

  @Column()
  isOpened: boolean;

  @Column({ type: 'time', nullable: true })
  timeStart: string | null;

  @Column({ type: 'time', nullable: true })
  timeEnd: string | null;
}

@Entity('producers')
export class Producer {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ length: 40 })
  producerName: string;

  @OneToMany(() => Goods, goods => goods.producer)
  goods: Goods[];
}

@Entity('goods')
export class Goods {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ length: 64 })
  name: string;

  @Column({ length: 64 })
  image: string;

  @Column()
  producerId: number;

  @Column()
  isInStock: boolean;

  @Column({ length: 1024, nullable: true })
  instruction: string;

  @Column({ length: 1024, nullable: true })
  description: string;

  @Column()
  isPrescriptionNeeded: boolean;

  @Column()
  price: number;

  @ManyToOne(() => Producer, producer => producer.goods)
  @JoinColumn({ name: 'producerId' })
  producer: Producer;

  @OneToMany(() => GoodsOrders, goodsOrder => goodsOrder.goods)
  goodsOrders: GoodsOrders[];

  @OneToMany(() => GoodsTag, goodsTag => goodsTag.goods)
  goodsTags: GoodsTag[];
}

@Entity('tags')
export class Tag {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ length: 20 })
  tagName: string;

  @OneToMany(() => GoodsTag, goodsTag => goodsTag.tag)
  goodsTags: GoodsTag[];
}

@Entity('goods_tags')
export class GoodsTag {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  goodsId: number;

  @Column()
  tagId: number;

  @ManyToOne(() => Goods, goods => goods.goodsTags)
  @JoinColumn({ name: 'goodsId' })
  goods: Goods;

  @ManyToOne(() => Tag, tag => tag.goodsTags)
  @JoinColumn({ name: 'tagId' })
  tag: Tag;
}

@Entity('orders')
export class Order {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ length: 30 })
  clientFIO: string;

  @Column({ length: 30, nullable: true })
  clientEmail: string;

  @Column({ length: 18 })
  clientPhone: string;

  @OneToMany(() => GoodsOrders, goodsOrder => goodsOrder.order, {
    onDelete: 'CASCADE'
  })
  goodsOrders: GoodsOrders[];
}

@Entity('goods_orders')
export class GoodsOrders {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  orderId: number;

  @Column()
  goodsId: number;

  @Column()
  quantity: number;

  @ManyToOne(() => Order, order => order.goodsOrders)
  @JoinColumn({ name: 'orderId' })
  order: Order;

  @ManyToOne(() => Goods, goods => goods.goodsOrders)
  @JoinColumn({ name: 'goodsId' })
  goods: Goods;
}

@Entity('announcements')
export class Announcement {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: 'timestamp' })
  dateTime: Date;

  @Column({ length: 64 })
  from: string;

  @Column({ length: 2048 })
  announce: string;
}

@Entity('permissions')
export class Permission {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ length: 20 })
  action: string;

  @ManyToMany(() => Role, role => role.permissions)
  roles: Role[];
}

@Entity('roles')
export class Role {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ length: 20 })
  name: string;

  @OneToMany(() => User, user => user.role)
  users: User[];

  @ManyToMany(() => Permission, permission => permission.roles)
  @JoinTable({
    name: 'role_permissions',
    joinColumn: { name: 'roleId', referencedColumnName: 'id' },
    inverseJoinColumn: { name: 'permissionId', referencedColumnName: 'id' }
  })
  permissions: Permission[];
}

@Entity('users')
export class User {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ length: 40, unique: true })
  login: string;

  @Column({ length: 40 })
  userName: string;

  @Column()
  roleId: number;

  @Column({ type: 'varbinary', length: 128 })
  passwordHash: string;

  @ManyToOne(() => Role, role => role.users)
  @JoinColumn({ name: 'roleId' })
  role: Role;
}

export const AppDataSource = new DataSource({
  type: 'mysql',
  host: process.env.DB_HOST,
  port: parseInt(process.env.DB_PORT || '3306'),
  username: process.env.DB_USER,
  password: process.env.DB_PASSWORD,
  database: process.env.DB_NAME,
  charset: 'utf8mb4',
  synchronize: false,
  logging: false,
  entities: [ Schedule, Producer, Tag, GoodsTag, GoodsOrders, Order, Goods, Announcement, User, Role, Permission ],
  migrations: [Pharmacy1776354362570],
});